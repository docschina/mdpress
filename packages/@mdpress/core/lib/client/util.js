import layoutComponents from '@internal/layout-components';
import pageComponents from '@internal/page-components';
import sandbox from '@internal/sandbox';
import Loadable from './lazy/Loadable';

/**
 * Create a cached version of a pure function.
 */
function cached (fn) {
  const cache = Object.create(null);
  // eslint-disable-next-line func-names
  return function cachedFn (str) {
    const hit = cache[str];
    // eslint-disable-next-line no-return-assign
    return hit || (cache[str] = fn(str));
  };
}

/**
 * Camelize a hyphen-delimited string.
 */
const camelizeRE = /-(\w)/g;
const camelize = cached(str => {
  return str.replace(camelizeRE, (_, c) => c ? c.toUpperCase() : '');
});

/**
 * Hyphenate a camelCase string.
 */
const hyphenateRE = /\B([A-Z])/g;
const hyphenate = cached(str => {
  return str.replace(hyphenateRE, '-$1').toLowerCase();
});

/**
 * Capitalize a string.
 */
const capitalize = cached(str => {
  return str.charAt(0).toUpperCase() + str.slice(1);
});

/**
 * This method was for securely getting React component when components
 * are named in different style.
 *
 * e.g. a component named `a-b` can be also getted by `AB`, It's the
 * same the other way round
 *
 * @param {function} getter a function of getting component by name
 * @param {string} name component's name
 * @returns {Component|AsyncComponent}
 */
export function getComponent (getter, name) {
  if (!name) return;
  if (getter(name)) return getter(name);

  const isKebabCase = name.includes('-');
  if (isKebabCase) return getter(capitalize(camelize(name)));

  return getter(capitalize(name)) || getter(hyphenate(name));
}

const asyncComponents = Object.assign({}, layoutComponents, pageComponents,sandbox);
const asyncComponentsGetter = name => asyncComponents[name];
const pageComponentsGetter = layout => pageComponents[layout];
const layoutComponentsGetter = layout => layoutComponents[layout];
const sandboxComponentsGetter = name => sandbox[name];
const globalComponentsGetter = name => Loadable.loadedMap[name];

export function getPageAsyncComponent (pageKey) {
  return getComponent(pageComponentsGetter, pageKey);
}

export function getLayoutAsyncComponent (layout) {
  return getComponent(layoutComponentsGetter, layout);
}

export function getSandboxAsyncComponent (name) {
  return getComponent(sandboxComponentsGetter, name);
}

export function getAsyncComponent (name) {
  return getComponent(asyncComponentsGetter, name);
}

export function getSyncComponent (name) {
  return getComponent(globalComponentsGetter, name);
}

export function findPageForPath (pages, path) {
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i];
    if (page.path === path) {
      return page;
    }
  }
  return {
    path: '',
    frontmatter: {}
  };
}

export function findPageByKey (pages, key) {
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i];
    if (page.key === key) {
      return page;
    }
  }
  return {
    path: '',
    frontmatter: {}
  };
}

/**
 * Normalize config.
 * This utility is mainly for plugin developers. For some
 * plugins that need internationalize the text. but it's
 * not recommenbded to let plugin care about to the internal
 * i18n implementation, so this utility was born.
 *
 *
 * Usage:
 *
 * import { normalizeConfig } from '@app/util'
 * export default {
 *   data () {
 *     return { config }
 *   }
 *   computed: {
 *     normalizedConfig() {
 *       return normalizeConfig(this, config)
 *     }
 *   }
 * }
 *
 *
 * e.g.
 *
 * Config: : 'Text'
 * Normalized Config: 'Text'
 *
 * Config: : { '/': 'Text', '/zh/': '文本' }
 * Normalized Config: 'Text' or '文本'
 *
 * @param {any} props
 * @param {any} rawConfig
 * @returns {any}
 */
export function normalizeConfig (props, rawConfig) {
  const { $localePath } = props;
  if (typeof rawConfig === 'object' && rawConfig[$localePath]) {
    return rawConfig[$localePath];
  }
  return rawConfig;
}

/**
 * Set global info in `window.__MDPRESS__` for debugging.
 *
 * @param {string}key
 * @param {any} value
 */
export function setGlobalInfo (key, value) {
  if (typeof window === 'undefined' || !window.__MDPRESS__) {
    return;
  }
  window.__MDPRESS__[key] = value;
}

function isNumber (v) {
  return typeof v === 'number';
}

export function normalizeOffset (dom) {
  let offset =
      dom.offset && typeof dom.offset === 'object'
        ? dom.offset
        : {};
  return {
    x: isNumber(offset.x) ? offset.x : 0,
    y: isNumber(offset.y) ? offset.y : 0
  };
}

export function getElementPosition (el, offset) {
  const docEl = document.documentElement;
  const docRect = docEl.getBoundingClientRect();
  const elRect = el.getBoundingClientRect();
  return {
    x: elRect.left - docRect.left - offset.x,
    y: elRect.top - docRect.top - offset.y
  };
}

export function getEnv() {
  return typeof window === 'undefined' ? 'server' : 'browser';
}