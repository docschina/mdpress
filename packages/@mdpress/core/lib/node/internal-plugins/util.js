const indexRE = /(^|.*\/)(index|readme)\.md$/i;
const extRE = /\.(js|md)$/;
exports.fileToComponentName = function (file) {
  let normalizedName = file
    .replace(/\/|\\|-/g, '_')
    .replace(extRE, '');
  if (exports.isIndexFile(file)) {
    normalizedName = normalizedName.replace(/readme$/i, 'index');
  }
  const pagePrefix = /\.md$/.test(file) ? 'page_' : '';
  return `${pagePrefix}${normalizedName}`;
};

exports.isIndexFile = function (file) {
  return indexRE.test(file);
};

exports.sort = function (arr) {
  return arr.sort((a, b) => {
    if (a < b) return -1;
    if (a > b) return 1;
    return 0;
  });
};

exports.genImportComponentCode = function (filename,key) {
  return `LazyLoadComponent(() => import(${JSON.stringify(filename)}),"${key}")`;
};

exports.commonImportCode = 'import { LazyLoadComponent } from \'@app/LazyLoader\';\n' +
    'import { getEnv } from \'@app/util\';\n' +
    'const isServer = getEnv() === "server";\n';