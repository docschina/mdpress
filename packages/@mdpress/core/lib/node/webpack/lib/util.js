'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.isJS = function (file) { return /\.js(\?[^.]+)?$/.test(file); };
exports.isCSS = function (file) { return /\.css(\?[^.]+)?$/.test(file); };
function createPromiseCallback() {
  var resolve;
  var reject;
  // tslint:disable-next-line variable-name
  var promise = new Promise(function (_resolve, _reject) {
    resolve = _resolve;
    reject = _reject;
  });
  var cb = function (err, res) {
    if (err) {
      return reject(err);
    }
    resolve(res || '');
  };
  return { promise: promise, cb: cb };
}
exports.createPromiseCallback = createPromiseCallback;
//# sourceMappingURL=util.js.map