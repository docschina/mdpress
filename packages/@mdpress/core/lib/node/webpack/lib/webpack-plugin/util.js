'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
  return (mod && mod.__esModule) ? mod : { 'default': mod };
};
Object.defineProperty(exports, '__esModule', { value: true });
var chalk_1 = __importDefault(require('chalk'));
var red = chalk_1.default.red, yellow = chalk_1.default.yellow;
var prefix = '[react-server-renderer-webpack-plugin]';
// tslint:disable-next-line:no-console
var warn = (exports.warn = function (msg) { return console.error(red(prefix + ' ' + msg + '\n')); });
// tslint:disable-next-line:no-console
var tip = (exports.tip = function (msg) { return console.log(yellow(prefix + ' ' + msg + '\n')); });
exports.validate = function (compiler) {
  if (compiler.options.target !== 'node') {
    warn('webpack config `target` should be "node".');
  }
  if (compiler.options.output &&
        compiler.options.output.libraryTarget !== 'commonjs2') {
    warn('webpack config `output.libraryTarget` should be "commonjs2".');
  }
  if (!compiler.options.externals) {
    tip('It is recommended to externalize dependencies in the server build for ' +
            'better build performance.');
  }
};
exports.onEmit = function (compiler, name, hook) {
  if (compiler.hooks) {
    // Webpack >= 4.0.0
    compiler.hooks.emit.tapAsync(name, hook);
  }
  else {
    // Webpack < 4.0.0
    compiler.plugin('emit', hook);
  }
};
var util_1 = require('../util');
exports.isJS = util_1.isJS;
exports.isCSS = util_1.isCSS;
//# sourceMappingURL=util.js.map