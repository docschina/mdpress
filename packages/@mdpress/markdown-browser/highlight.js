import prism from 'prismjs';
import 'prismjs/components/prism-markdown.min';
import 'prismjs/components/prism-jsx.min';

prism.manual = true;

function wrap (code, lang) {
  return `<pre class="language-${lang}"><code>${code}</code></pre>`;
}

export default (str = '', lang) => {
  if (!lang) {
    lang = 'text';
  }
  const rawLang = lang;
  if (lang === 'vue' || lang === 'html') {
    lang = 'markup';
  }
  if (lang === 'md') {
    lang = 'markdown';
  }
  if (prism.languages[lang]) {
    const code = prism.highlight(str, prism.languages[lang], lang);
    return wrap(code, rawLang);
  }
  const code = prism.highlight(str, prism.languages['markdown'], 'markdown');
  return wrap(code, 'text');
};
