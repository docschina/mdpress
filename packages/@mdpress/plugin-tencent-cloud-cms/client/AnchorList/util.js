export function genAnchors(dom) {
  if (!dom || !dom.children) return [];
  const children = Array.from(dom.children);
  const tree = children.reduce((res,child) => {
    let nodeName = child.nodeName.toLowerCase();
    if (nodeName.match(/h[1-6]/)) {
      return res.concat([{
        name: child.innerText.replace(/^#[\n\s]/,''),
        tag: child.nodeName,
        level: +nodeName.replace('h',''),
        id: child.id
      }]);
    }

    return res;

  },[]);

  return format(tree);
}

function format(treeArray) {
  let result = [];
  treeArray.forEach(function (item) {
    if (!result.length) {
      result.push(item);
      return;
    }
    let prevItem = last(result);
    let level = item.level;
    let prevLevel = prevItem.level;
    // 比较级别大小
    let isSmall = level > prevLevel;
    if (isSmall) {
      for (let i = prevLevel + 1; i <= 3; i++) {
        let children = prevItem.children;
        if (!children) {
          prevItem.children = [item];
          break;
        }
        // 重置 lastItem 为 children 的最后一个 item
        prevItem = last(children);
        // item level 小于或等于 lastItem level 都视为与 children 同级
        if (item.level <= prevItem.level) {
          children.push(item);
          break;
        }
      }
    }
    else {
      result.push(item);
    }
  });
  return result;
}

function last(array) {
  return array[array.length - 1];
}
