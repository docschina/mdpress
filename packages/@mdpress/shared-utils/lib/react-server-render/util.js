"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPromiseCallback = exports.isCSS = exports.isJS = void 0;
exports.isJS = (file) => /\.js(\?[^.]+)?$/.test(file);
exports.isCSS = (file) => /\.css(\?[^.]+)?$/.test(file);
function createPromiseCallback() {
    let resolve;
    let reject;
    // tslint:disable-next-line variable-name
    const promise = new Promise((_resolve, _reject) => {
        resolve = _resolve;
        reject = _reject;
    });
    const cb = (err, res) => {
        if (err) {
            return reject(err);
        }
        resolve(res || '');
    };
    return { promise, cb };
}
exports.createPromiseCallback = createPromiseCallback;
