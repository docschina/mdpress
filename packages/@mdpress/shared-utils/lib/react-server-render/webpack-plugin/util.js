"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.onEmit = exports.validate = void 0;
const chalk_1 = __importDefault(require("chalk"));
const { red, yellow } = chalk_1.default;
const prefix = '[react-server-renderer-webpack-plugin]';
// tslint:disable-next-line:no-console
const warn = (exports.warn = msg => console.error(red(`${prefix} ${msg}\n`)));
// tslint:disable-next-line:no-console
const tip = (exports.tip = msg => console.log(yellow(`${prefix} ${msg}\n`)));
exports.validate = compiler => {
    if (compiler.options.target !== 'node') {
        warn('webpack config `target` should be "node".');
    }
    if (compiler.options.output &&
        compiler.options.output.libraryTarget !== 'commonjs2') {
        warn('webpack config `output.libraryTarget` should be "commonjs2".');
    }
    if (!compiler.options.externals) {
        tip('It is recommended to externalize dependencies in the server build for ' +
            'better build performance.');
    }
};
exports.onEmit = (compiler, name, hook) => {
    if (compiler.hooks) {
        // Webpack >= 4.0.0
        compiler.hooks.emit.tapAsync(name, hook);
    }
    else {
        // Webpack < 4.0.0
        compiler.plugin('emit', hook);
    }
};
var util_1 = require("../util");
Object.defineProperty(exports, "isJS", { enumerable: true, get: function () { return util_1.isJS; } });
Object.defineProperty(exports, "isCSS", { enumerable: true, get: function () { return util_1.isCSS; } });
