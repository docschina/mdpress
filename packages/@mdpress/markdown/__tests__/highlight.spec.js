import { getFragment } from '@mdpress/test-utils';
import { Md } from './util';
import highlight from '../lib/highlight.js';

const md = Md();
const mdH = Md().set({ highlight });

describe('highlight', () => {
  test('should highlight code', () => {
    const input = getFragment(__dirname, 'code.md');
    const output1 = md.render(input);
    const output2 = mdH.render(input);
    expect(output1 === output2).toBe(false);
    expect(output2).toMatchSnapshot();
  });
});
