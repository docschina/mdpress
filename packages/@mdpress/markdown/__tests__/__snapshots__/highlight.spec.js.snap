// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlight should highlight code 1`] = `
"<pre v-pre class=\\"language-js\\"><code><span class=\\"token keyword\\">const</span> array <span class=\\"token operator\\">=</span> <span class=\\"token keyword\\">new</span> <span class=\\"token class-name\\">Array</span><span class=\\"token punctuation\\">(</span><span class=\\"token punctuation\\">)</span><span class=\\"token punctuation\\">;</span>
</code></pre>
"
`;
